sys/include/xtimer.h

```diff
@@ -0,0 +1,442 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup  sys_xtimer Timers
+ * @ingroup   sys
+ * @brief     Provides a high level timer module to register
+ *            timers, get current system time, and let a thread sleep for
+ *            a certain amount of time.
+ *
+ * The implementation takes one low-level timer that is supposed to run at 1MHz
+ * speed and multiplexes it.
+ *
+ * Insertion and removal of timers has O(n) complexity with (n) being the
+ * number of active timers.  The reason for this is that multiplexing is
```

Comment:
> Usually real-times are defined by meeting certain deadlines. Hence, the WCET is what matters.

You state this as if it was a new insight. :)

Again, we could limit the amount of active timers in order to have a
maximum upper bound of cycles needed for insertion or removal of timers.

*but*

1. it would mean that currently never-fail functions might fail, forcing
any user of xtimers to deal with a failing timer, making the overall
codebase larger and more complex

2. users of xtimer with realtime requirements will need to make sure
that the system is not using the maximum amount of timers, as otherwise,
our now-deterministic timer might indeterministically fail. So while now
setting of the timer is deterministic, the application timing is not, as
depending on the amount of timers in the system, a failed timer must be
dealt with. While this can be dealt with in a more controlled manner, it
is not trivial to still keep timing constraints at a precision where the
number of timers make a significant difference.

IMHO, the gain of limiting the number of timers in order to get an upper
bound of execution time is little to nonexistant and doesn't justify the
increased complexity. It might, though, satisfy academic WCET requirements.